### PR TITLE
Do not renew sync-id if all shards are sealed

### LINKED ISSUE #27838: Index with no changes gets new `sync_id` when becoming inactive.
**Elasticsearch version** (`bin/elasticsearch --version`):

Version: 5.5.2, Build: b2f0c09/2017-08-14T12:33:14.154Z, JVM: 1.8.0_144

**Plugins installed**: []

**JVM version** (`java -version`):

java version "1.8.0_144"
Java(TM) SE Runtime Environment (build 1.8.0_144-b01)
Java HotSpot(TM) 64-Bit Server VM (build 25.144-b01, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):

Linux xxx 3.13.0-33-generic #58-Ubuntu SMP Tue Jul 29 16:45:05 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

The problem is that restarting a node takes very long. It takes several hours to wait for a cluster to become *green* after restarting just one single node, even if there are no indexing and sync flush was successfully executed. The restarted node should be able to recover from local files since they are up to date, but it doesn't because the `sync_id` doesn't match.

Expected behavior: After shutting down indexing and performing a successful synced flush it is expected that `sync_id`s stays the same.

Actual behavior: After shutting down indexing and performing a successful synced flush the `sync_id` changes when the index becomes *inactive*.

**Steps to reproduce**:

```shell
# Create replicated index with single shard
curl -XPUT localhost:9200/testindex -d '{"settings":{"index":{
  "number_of_shards" : 1,
  "number_of_replicas" : 1
}}}'

# Index one document
curl -XPUT localhost:9200/testindex/doc/1 -d '{}'

# Perform a synced flush
curl -XPOST localhost:9200/_flush/synced

# Note the sync_id (if jq is not installed; navigate JSON correspondingly)
curl localhost:9200/_stats?level=shards | jq '.indices.testindex.shards["0"][].commit.user_data.sync_id'

# wait 5-6 minutes

# The sync_id has changed:
curl localhost:9200/_stats?level=shards | jq '.indices.testindex.shards["0"][].commit.user_data.sync_id'
```

**Provide logs (if relevant)**:

With debug logging enabled, something like this is printed:
```
[2017-12-15T11:30:21,439][DEBUG][o.e.i.e.Engine           ] [xxx] [testindex][0] successfully sync committed. sync id [AWBZ8H59yP3Rl7MoZfSx].
[2017-12-15T11:30:22,312][DEBUG][o.e.i.s.IndexShard       ] [xxx] [testindex][0] shard is now inactive
```

### PR DESCRIPTION
**Pull request by dnhatn**
Today the synced-flush always issues a new sync-id even though all
shards haven't been changed since the last seal. This causes active
shards to have different a sync-id from offline shards even though all
were sealed and no writes since then.

This commit adjusts not to renew sync-id if all active shards are sealed
with the same sync-id.

Closes #27838

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java [hunk 1]
```diff
@@ -280,4 +281,35 @@ public void testSyncedFlushSkipOutOfSyncReplicas() throws Exception {
...
+        for (int i = 0; i < numDocs; i++) {
+            index("test", "doc", Integer.toString(i));
+        }
+        final ShardsSyncedFlushResult firstSeal = SyncedFlushUtil.attemptSyncedFlush(internalCluster(), shardId);
+        assertThat(firstSeal.successfulShards(), equalTo(numberOfReplicas + 1));
+        // Do not renew synced-flush
+        final ShardsSyncedFlushResult secondSeal = SyncedFlushUtil.attemptSyncedFlush(internalCluster(), shardId);
+        assertThat(secondSeal.successfulShards(), equalTo(numberOfReplicas + 1));
+        assertThat(secondSeal.syncId(), equalTo(firstSeal.syncId()));
+        // Shards were updated, renew synced flush
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/indices/flush/FlushIT.java]: can we inject/remove the marker from one of the shards and see that the synced flush is re-applied?

**Review by bleskes — APPROVED** [review]
LGTM. I like how contained it is (i.e. not trying to cover all edge cases where some of the shards may have another seal id in place or trying to reuse the primary's sync id marker). I left some simple ask for testing. No need for another cycle

**Comment by dnhatn** [comment]
Thanks @bleskes for reviewing.

**Comment by DaveCTurner** [comment]
@dnhatn I backed out the backport to the 5.6 branch because it seems to be the source of the test failures in #29162.

**Pull request merged**